src/main.rs

```diff
@@ -0,0 +1,546 @@
+        translation.y += velocity.y;
+    }
+}
+
+// Calculates new velocity vectors after collision.
+// Using some math formulas from the internet.
+// Inspired with: https://stackoverflow.com/questions/345838/ball-to-ball-collision-detection-and-handling
+fn handle_collision(
+    velocity1: &mut Velocity,
+    velocity2: &mut Velocity,
+    transform_red: &mut Transform,
+    transform_blue: &mut Transform,
+    radius1: f32,
+    radius2: f32,
+) {
+    let delta = (transform_red.translation - transform_blue.translation).truncate();
+    let players_distance = transform_red
+        .translation
+        .distance(transform_blue.translation);
+    let d = players_distance;
+    let multiplier = (-d + radius1 + radius2) / d;
+    let delta_x = delta.x * multiplier;
+    let delta_y = delta.y * multiplier;
+    let mtd = Vec2::new(delta_x, delta_y);
+
+    let im1 = 1.;
+    let im2 = 1.;
+
+    transform_red.translation.x += mtd[0] * (im1 / (im1 + im2));
+    transform_red.translation.y += mtd[1] * (im1 / (im1 + im2));
+
+    transform_blue.translation.x -= mtd[0] * (im2 / (im1 + im2));
+    transform_blue.translation.y -= mtd[1] * (im2 / (im1 + im2));
+
+    let v = Vec2::new(velocity1.x - velocity2.x, velocity1.y - velocity2.y);
+    let vn = v.dot(mtd.normalize());
+
+    if vn > 0.0 {
+        return;
+    }
+
+    let i = (-1.5 * vn) / (im1 + im2);
+    let impulse = mtd.normalize() * i;
+
+    velocity1.x += impulse[0] * im1;
+    velocity1.y += impulse[1] * im1;
+
+    velocity2.x -= impulse[0] * im2;
+    velocity2.y -= impulse[1] * im2;
+}
+
+// Detects collision between red player and the ball.
+fn collision_system_red(
+    mut query_red: Query<(&mut Velocity, &mut Transform, &PlayerRed), Without<Ball>>,
+    mut query_ball: Query<(&mut Velocity, &mut Transform, &Ball, Without<PlayerRed>)>,
+    kb: Res<Input<KeyCode>>,
+) {
+    let (mut velocity_red, mut transform_red, _) = query_red.iter_mut().next().unwrap();
+    let (mut velocity_ball, mut transform_ball, _, _) = query_ball.iter_mut().next().unwrap();
+
+    let player_ball_distance = transform_red
+        .translation
+        .distance(transform_ball.translation);
+    if player_ball_distance < PLAYER_RADIUS + BALL_RADIUS {
+        // If space pressed, shoot the ball
+        if kb.pressed(KeyCode::Space) {
+            let diff_x = transform_red.translation.x - transform_ball.translation.x;
+            let diff_y = transform_red.translation.y - transform_ball.translation.y;
+            let angle = diff_y.atan2(diff_x);
+            velocity_ball.y += -5.0 * angle.sin();
+            velocity_ball.x += -5.0 * angle.cos();
+        }
+        handle_collision(
+            &mut velocity_red,
+            &mut velocity_ball,
+            &mut transform_red,
+            &mut transform_ball,
+            PLAYER_RADIUS,
+            BALL_RADIUS,
+        );
+    }
+}
+
+// Detects collision between blue player and the ball.
+fn collision_system_blue(
```

Comment:
ponieważ ten system też jest analogiczny, to proponowałbym wyciągniecie wspólnej logiki do funkcji